segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,186 @@
+
+                long version = ((Long) txnData.getDbObject()).longValue();
+                if (null == txnData.getValue()) {
+                    val toDelete = TableKey.versioned(new ByteArraySegment(txnData.getKey().getBytes()),
+                            TableKey.NO_VERSION);
+                    keysToDelete.add(toDelete);
+                    deletedKeyToTxnDataMap.put(toDelete, txnData);
+                }
+
+                val arraySegment = serializer.serialize(txnData);
+
+                TableEntry tableEntry = TableEntry.versioned(
+                        new ByteArraySegment(txnData.getKey().getBytes()),
+                        arraySegment,
+                        version);
+                entryToTxnDataMap.put(tableEntry, txnData);
+                toUpdate.add(tableEntry);
+            }
+
+            // Now put uploaded keys.
+            List<Long> ret = this.tableStore.put(tableName, toUpdate, timeout).get();
+
+            // Update versions.
+            int i = 0;
+            for (TableEntry tableEntry : toUpdate) {
+                entryToTxnDataMap.get(tableEntry).setDbObject(ret.get(i));
+                i++;
+            }
+
+            // Delete deleted keys.
+            this.tableStore.remove(tableName, keysToDelete, timeout).get();
+            for (val deletedKey : keysToDelete) {
+                deletedKeyToTxnDataMap.get(deletedKey).setDbObject(TableKey.NOT_EXISTS);
+            }
+        } catch (RuntimeException e) {
+            throw e; // To make spotbugs happy.
+        } catch (java.util.concurrent.ExecutionException e) {
+            handleException(e.getCause());
+            return;
+        } catch (Exception e) {
+            handleException(e);
+            return;
+        }
+
+    }
+
+    private void handleException(Throwable e) throws StorageMetadataException {
+        if (e instanceof DataLogWriterNotPrimaryException) {
+            throw new StorageMetadataWritesFencedOutException("Transaction failed. Writer fenced off", e);
+        }
+        if (e instanceof BadKeyVersionException) {
+            throw new StorageMetadataVersionMismatchException("Transaction failed. Version Mismatch.", e);
+        }
+        if (e.getCause() != null) {
+            if (e.getCause().getCause() instanceof BadKeyVersionException) {
+                throw new StorageMetadataWritesFencedOutException("Transaction writer is fenced off.", e);
+            }
+            if (e.getCause().getCause() instanceof DataLogWriterNotPrimaryException) {
+                throw new StorageMetadataVersionMismatchException("Transaction failed. Writer fenced off", e);
+            }
+        } else {
+            log.debug("e.getCause()=null", e);
+        }
+        throw new StorageMetadataException("Transaction failed", e);
+    }
+
+    private void ensureInitialized() {
+        if (!isTableInitialized.get()) {
+            try {
+                this.tableStore.createSegment(tableName, timeout).join();
```

Comment:
Same comment here about making this be async.

Furthermore, you use `join` here and above you use `get`.

Reply:
same as above.